### PR TITLE
Support sync batch norm.

### PR DESCRIPTION
**Pull request by qingqing01**
1. Now only support GPU, CPU will be add in nexted PR.
2. Use ncclAllReduce to sync E(x) and E(x^2) on multi-gpus in one machine.
~~3. Add `ncclComm_t` in DeviceContext and initialized by init.cc~~
4. The unit testing is to compare the forward outputs and backward outputs with batch_norm on one GPU.

```python
build_strategy = fluid.BuildStrategy()
build_strategy.sync_batch_norm = True
binary = fluid.compiler.CompiledProgram(tp).with_data_parallel(
        loss_name=loss_mean.name,
        build_strategy=build_strategy)
```

### CODE REVIEW

File: paddle/fluid/operators/batch_norm_op.h [hunk 1]
```diff
@@ -35,17 +38,84 @@ template <typename T>
...
 };
 
 template <typename DeviceContext, typename T>
 class BatchNormGradKernel : public framework::OpKernel<T> {
  public:
-  void Compute(const framework::ExecutionContext& ctx) const override;
+  void Compute(const framework::ExecutionContext &ctx) const override;
 };
 
+inline void ExtractNCWHD(const framework::DDim &dims,
```
Comment (heavengate) [inline paddle/fluid/operators/batch_norm_op.h:99]: static ？

File: paddle/fluid/operators/sync_batch_norm_op.cc [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline paddle/fluid/operators/sync_batch_norm_op.cc]: 2016 -> 2019

File: paddle/fluid/operators/sync_batch_norm_op.cu [hunk 3]
```diff
@@ -0,0 +1,452 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline paddle/fluid/operators/sync_batch_norm_op.cu]: 2016 -> 2019

File: python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                    out = fluid.layers.reduce_sum(sigmoid)
+                    out = out / core.get_cuda_device_count()
+                if not only_forward:
+                    sgd_opt = fluid.optimizer.SGD(learning_rate=0.0)
+                    sgd_opt.backward(out)
+        return main, startup, [out, conv, bn]
+
+    def compare(self, place, layout, only_forward):
+        seed = 10
+        #np.random.seed(5)
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py]: remove this line ?

File: python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py]: 2018 -> 2019

**Review by chengduoZH — COMMENTED** [review]
After a general look, I think this design may be a bit inappropriate, first, we should not put nccl_comm_ in device_context, because, in some case, the user may want to create two different ParallelExecutor with different GPU to run two different models, obviously, the current design doesn't allow this. Second, the sync_batch_norm_op may induce the program slow, in this case, why not combine a sync_batch_norm_op with some ops? Currently, @typhoonzero is designing [`Collective ops`](https://github.com/PaddlePaddle/Paddle/pull/15572), maybe you can use  collective op and other ops to compose sync_batch_norm_op.

File: paddle/fluid/operators/sync_batch_norm_op.cu [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+                           framework::DataLayout::kNCHW><<<C, 512, 0, stream>>>(
+          dy_d, x_d, saved_mean, N, H * W * D, C, stats);
+    } else {
+      KeBackwardLocalStats<T, 512,
+                           framework::DataLayout::kNHWC><<<C, 512, 0, stream>>>(
+          dy_d, x_d, saved_mean, N, H * W * D, C, stats);
+    }
+    int dtype = platform::ToNCCLDataType(x->type());
+    // In-place operation
+    PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
```
Comment (panyx0718) [inline paddle/fluid/operators/sync_batch_norm_op.cu:401]: can training hung if the op order on different device is different?

Comment (chengduoZH) [reply paddle/fluid/operators/sync_batch_norm_op.cu:401]: Maybe it will be hung if there is more than one sync_batch_norm_op.

File: paddle/fluid/platform/init.cc [hunk 7]
```diff
@@ -132,14 +138,36 @@ void InitDevices(bool init_p2p, const std::vector<int> devices) {
...
+#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32) && \
+    !defined(PADDLE_ON_INFERENCE)
+  std::vector<platform::Place> cuda_devs;
+  for (auto &p : places) {
+    if (platform::is_gpu_place(p)) {
+      cuda_devs.push_back(p);
+    }
+  }
+  std::unique_ptr<platform::NCCLContextMap> nccl_ctxs;
+  nccl_ctxs.reset(new platform::NCCLContextMap(cuda_devs));
```
Comment (typhoonzero) [inline paddle/fluid/platform/init.cc]: Maybe better to initialize at `ParallelExecutor` to be compatible with distributed mode, which needs to pass global number of devices in the cluster. In `init.cc` it can only get local device count.

File: paddle/fluid/operators/sync_batch_norm_op.cu [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+                                     const double epsilon, T *sv_mean_data,
+                                     T *sv_inv_var_data, T *moving_means,
+                                     T *moving_variances) {
+  // sync stats across multi-devices
+  int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  int stride = blockDim.x * gridDim.x;
+  for (int i = gid; i < C; i += stride) {
+    T mean = means[i] / (*num_dev);
+    T var = variances[i] / (*num_dev);
+    var = var - mean * mean;
```
Comment (typhoonzero) [inline paddle/fluid/operators/sync_batch_norm_op.cu:78]:
```
var = (sum((x - mean) ** 2)) / batch_size
var = (sum(x**2) - sum(2*x*mean) + sum(mean)**2) / batch_size
```
???

Comment (qingqing01) [reply paddle/fluid/operators/sync_batch_norm_op.cu:78]: @typhoonzero Here, use v = E(x^2) - E(x)^2 to calculate variance

**Comment by qingqing01** [comment]
>  Currently, @typhoonzero is designing Collective ops, maybe you can use collective op and other ops to compose sync_batch_norm_op.

@chengduoZH @typhoonzero 

If  use ops to combine:

1. Need to calculate local E(x) and  E(x2) by ops
2. Then sync E(x) and  E(x2) by collective allreduce op.
3. Then y = scale * (x-mean)/sqrt(var + epsion) + bias

This needs to split into many ops.

Use sync BN due to the memory and small batch size, if use many ops to combine sync_bn, more memory is used.  I prefer to one op to implement sync_bn, and I will rewrite the backward of cudnn_batch_norm and sync_batch_norm to make the BN in-place in next PR.

File: paddle/fluid/operators/sync_batch_norm_op.cu (hunk 8 above)
Comment (typhoonzero) [reply paddle/fluid/operators/sync_batch_norm_op.cu:78]: I see, thanks.

**Comment by typhoonzero** [comment]
@qingqing01 @chengduoZH After walking through this implementation, and discuss with @qingqing01 , I quite agree with current "big op" implement, it should be effective.

On the other hand, if we want to assemble things out of several operators and allreduce_op, it does not save lines of code except we can assemble the sync_bn layer using basic operators like reductions, element-wise ops, unary ops like tensorflow, if we do so, we must be able to fuse these operations automatically to achieve same performance.

So, LGTM for current design.

File: paddle/fluid/operators/sync_batch_norm_op.cc (hunk 2 above)
Comment (qingqing01) [reply paddle/fluid/operators/sync_batch_norm_op.cc]: Done.

File: paddle/fluid/operators/sync_batch_norm_op.cu (hunk 3 above)
Comment (qingqing01) [reply paddle/fluid/operators/sync_batch_norm_op.cu]: Done. Thanks!

File: python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py (hunk 4 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py]: Done. Thanks!

File: python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py (hunk 5 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/unittests/test_sync_batch_norm_op.py]: Done.

File: paddle/fluid/platform/init.cc (hunk 7 above)
Comment (qingqing01) [reply paddle/fluid/platform/init.cc]: Done. Thanks!

**Comment by qingqing01** [comment]
图片

@shanyi15 Please help to review Doc. Thanks!

**Comment by shanyi15** [comment]
Review result: 
另外 variance首字母有的是大写、有的小写，请统一
image

**Comment by qingqing01** [comment]
图片

图片

@shanyi Update.

**Review by shanyi15 — DISMISSED** [review]
LG comments

**Comment by qingqing01** [comment]
图片

@shanyi15 Update doc again.

**Comment by qingqing01** [comment]
I have tested on DeepLab v3+ by sync_bn:

```
step 42870, loss: 0.317351, step_time_cost: 1.188
step 42880, loss: 0.154967, step_time_cost: 1.271
step 42890, loss: 0.239504, step_time_cost: 1.260
step 42900, loss: 0.189635, step_time_cost: 1.218
Model is saved to output
step 42910, loss: 0.125752, step_time_cost: 1.216
step 42920, loss: 0.185016, step_time_cost: 1.265
step 42930, loss: 0.181721, step_time_cost: 1.250
```

@chengduoZH @typhoonzero @panyx0718 Please help to review.

**Comment by shanyi15** [comment]
看到基于上次修改,增加了红框处,请问还有其他diff吗
image

**Review by shanyi15 — DISMISSED** [review]
LG comments

File: paddle/fluid/operators/batch_norm_op.cc [hunk 9]
```diff
@@ -642,10 +624,10 @@ class BatchNormGradInplaceInToOut : public framework::InplaceInToOut {
 
 namespace ops = paddle::operators;
 REGISTER_OPERATOR(batch_norm, ops::BatchNormOp, ops::BatchNormOpMaker,
-                  ops::BatchNormOpInferVarType, ops::BatchNormGradMaker,
-                  ops::BatchNormInplaceInToOut);
-REGISTER_OPERATOR(batch_norm_grad, ops::BatchNormGradOp,
-                  ops::BatchNormGradInplaceInToOut);
+                  ops::BatchNormOpInferVarType, ops::BatchNormGradMaker)
+// ops::BatchNormInplaceInToOut);
```
Comment (panyx0718) [inline paddle/fluid/operators/batch_norm_op.cc:628]: why change here?

File: paddle/fluid/operators/sync_batch_norm_op.cu (hunk 6 above)
Comment (panyx0718) [reply paddle/fluid/operators/sync_batch_norm_op.cu:401]: how to solve this problem?

File: paddle/fluid/operators/batch_norm_op.cc (hunk 9 above)
Comment (qingqing01) [reply paddle/fluid/operators/batch_norm_op.cc:628]: See the `cudnnBatchNormalizationBackward`,  x is the input. So cuDNN BatchNorm does not support in-place, I think there is bug when enable in-place pass in CNN model.  And @DengKaiPeng enable BN-inplace  in vedio classfication, there  is problem in the training.     And as say above  " I will rewrite the backward  to make the BN in-place in next PR."

```
      CUDNN_ENFORCE(platform::dynload::cudnnBatchNormalizationBackward(
          dev_ctx.cudnn_handle(), mode_, CudnnDataType<T>::kOne(),
          CudnnDataType<T>::kZero(), CudnnDataType<T>::kOne(),
          CudnnDataType<T>::kZero(), data_desc_, x->template data<T>(),
          data_desc_, d_y->template data<T>(), data_desc_,
          d_x->template mutable_data<T>(ctx.GetPlace()), bn_param_desc_,
          scale->template data<BatchNormParamType<T>>(),
          d_scale->template mutable_data<BatchNormParamType<T>>(ctx.GetPlace()),
          d_bias->template mutable_data<BatchNormParamType<T>>(ctx.GetPlace()),
epsilon, saved_mean_data, saved_var_data));
```

File: paddle/fluid/operators/sync_batch_norm_op.cu (hunk 6 above)
Comment (qingqing01) [reply paddle/fluid/operators/sync_batch_norm_op.cu:401]: @panyx0718  

tell user to set `build_strategy.enable_sequential_execution = True` in the sync_batch_norm doc. And I test it in DeepLabv3+ on v100, the traning doest not  hung.

File: paddle/fluid/framework/parallel_executor.cc [hunk 10]
```diff
@@ -250,6 +252,17 @@ ParallelExecutor::ParallelExecutor(
     member_->nccl_ctxs_.reset(new platform::NCCLContextMap(
         member_->places_, nccl_id, build_strategy.num_trainers_,
         build_strategy.trainer_id_));
+
+    auto dev_nccl_ctxs = new platform::NCCLContextMap(member_->places_);
+    // initialize device context's nccl comm, so that all operators
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc]: It seems that line257 is incomplete.

File: paddle/fluid/platform/device_context.cc [hunk 11]
```diff
@@ -53,12 +53,7 @@ inline void EmplaceDeviceContext(
 DeviceContextPool::DeviceContextPool(
     const std::vector<platform::Place>& places) {
   PADDLE_ENFORCE_GT(places.size(), 0);
-  std::set<Place> set;
   for (auto& p : places) {
-    set.insert(p);
-  }
-
-  for (auto& p : set) {
```
Comment (chengduoZH) [inline paddle/fluid/platform/device_context.cc]: Why you remove the above code?

File: paddle/fluid/framework/parallel_executor.cc [hunk 12]
```diff
@@ -251,6 +253,17 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
     member_->nccl_ctxs_.reset(new platform::NCCLContextMap(
         member_->places_, nccl_id, build_strategy.num_trainers_,
         build_strategy.trainer_id_));
+
+    auto dev_nccl_ctxs = new platform::NCCLContextMap(member_->places_);
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc]: The `dev_nccl_ctxs ` is not deleted.

Comment (qingqing01) [reply paddle/fluid/framework/parallel_executor.cc]: Done

File: paddle/fluid/platform/device_context.cc (hunk 11 above)
Comment (qingqing01) [reply paddle/fluid/platform/device_context.cc]: Remove outside, but revert now.

File: paddle/fluid/framework/parallel_executor.cc (hunk 10 above)
Comment (qingqing01) [reply paddle/fluid/framework/parallel_executor.cc]: Done

File: paddle/fluid/framework/parallel_executor.cc (hunk 12 above)
Comment (qingqing01) [reply paddle/fluid/framework/parallel_executor.cc]: Done

**Review by chengduoZH — DISMISSED** [review]

**Review by shanyi15 — APPROVED** [review]

**Review by chengduoZH — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
